src/plot_api/plot_api.js

```diff
@@ -2208,6 +2195,356 @@ Plotly.update = function update(gd, traceUpdate, layoutUpdate, _traces) {
     });
 };
 
+/**
+ * Plotly.react:
+ * A plot/update method that takes the full plot state (same API as plot/newPlot)
+ * and diffs to determine the minimal update pathway
+ *
+ * @param {string id or DOM element} gd
+ *      the id or DOM element of the graph container div
+ * @param {array of objects} data
+ *      array of traces, containing the data and display information for each trace
+ * @param {object} layout
+ *      object describing the overall display of the plot,
+ *      all the stuff that doesn't pertain to any individual trace
+ * @param {object} config
+ *      configuration options (see ./plot_config.js for more info)
+ *
+ * OR
+ *
+ * @param {string id or DOM element} gd
+ *      the id or DOM element of the graph container div
+ * @param {object} figure
+ *      object containing `data`, `layout`, `config`, and `frames` members
+ *
+ */
+Plotly.react = function(gd, data, layout, config) {
+    var frames, plotDone;
+
+    function addFrames() { return Plotly.addFrames(gd, frames); }
+
+    gd = Lib.getGraphDiv(gd);
+
+    var oldFullData = gd._fullData;
+    var oldFullLayout = gd._fullLayout;
+
+    // you can use this as the initial draw as well as to update
+    if(!Lib.isPlotDiv(gd) || !oldFullData || !oldFullLayout) {
+        plotDone = Plotly.newPlot(gd, data, layout, config);
+    }
+    else {
+
+        if(Lib.isPlainObject(data)) {
+            var obj = data;
+            data = obj.data;
+            layout = obj.layout;
+            config = obj.config;
+            frames = obj.frames;
+        }
+
+        var configChanged = false;
+        // assume that if there's a config at all, we're reacting to it too,
+        // and completely replace the previous config
+        if(config) {
+            var oldConfig = Lib.extendDeep({}, gd._context);
+            gd._context = undefined;
+            setPlotContext(gd, config);
+            configChanged = diffConfig(oldConfig, gd._context);
+        }
+
+        gd.data = data || [];
+        helpers.cleanData(gd.data, []);
+        gd.layout = layout || {};
+        helpers.cleanLayout(gd.layout);
```

Comment:
I'm a little ambivalent about this - there's part of `cleanData`, fillling in uids, that seems like it may not even belong to `cleanData` but anyway we need to do it. But the rest, do we need to be cleaning obsolete usages on every `Plotly.react` call? These routines mutate the inputs so after the first call the user shouldn't have any of them left; that said users can always provide totally new data/layout...

Reply:
I agree, the `uid` generation routine should be taken out. It has no place in `cleanData`. 

In `v2`, I'm thinking `cleanData` and `cleanLayout` should be called only under a special config flag (e.g. `compatibility: true`, or a more granular `compatibilityLevel: 'v1'`) whereas assigning uid to traces will always have to be done.

cc https://github.com/plotly/plotly.js/issues/2007